### PR TITLE
[SPARK-39253][DOCS][PYTHON] Improve PySpark API reference to be more readable

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR proposes to improve the PySpark API reference page to be more readable,

So far, the PySpark documentation especially ["Spark SQL" part](https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql.html#) is not-well organized so it's a bit uncomfortable to be read as below:

Screen Shot 2022-05-24 at 1 47 18 PM

For example, whereas [pandas API on Spark reference page](https://spark.apache.org/docs/latest/api/python/reference/pyspark.pandas/index.html) is relatively well-organized as below:

Screen Shot 2022-05-24 at 1 47 30 PM

### Why are the changes needed?

The improvement of document readability will also improve the usability for PySpark.

### Does this PR introduce _any_ user-facing change?

Yes, now the documentation is categorized by its class or their own purpose more clearly as below:

Screen Shot 2022-05-24 at 1 50 23 PM

### How was this patch tested?

The CI already include the doc build, so the existing test should cover.

### CODE REVIEW

File: python/docs/source/reference/pyspark.sql/core_classes.rst [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    Catalog
+    DataFrame
+    Column
+    Observation
+    Row
+    GroupedData
+    PandasCogroupedOps
+    DataFrameNaFunctions
+    DataFrameStatFunctions
+    Window
```
Comment (HyukjinKwon) [inline python/docs/source/reference/pyspark.sql/core_classes.rst:37]: Can we have DataFrameReader DataFrameWriter?

Comment (itholic) [reply python/docs/source/reference/pyspark.sql/core_classes.rst:37]: It's included in I/O, do we want to have separated category ?

File: python/docs/source/reference/pyspark.sql/functions.rst [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+=========
+Functions
+=========
+
+.. currentmodule:: pyspark.sql.functions
+
+.. autosummary::
+    :toctree: api/
+
+    abs
```
Comment (HyukjinKwon) [inline python/docs/source/reference/pyspark.sql/functions.rst:28]: Should we maybe have a sub-group like we do in Pandas API on Spark?  e.g., Aggregate Function, Higher Order functions. See also https://github.com/apache/spark/blob/master/R/pkg/R/functions.R.

**Review by HyukjinKwon — COMMENTED** [review]
Does the URL change? We might need to keep the mapping between old URL and new URL. If that's complicated, we can do it in a followup.

File: python/docs/source/reference/pyspark.sql/core_classes.rst (hunk 1 above)
Comment (HyukjinKwon) [reply python/docs/source/reference/pyspark.sql/core_classes.rst:37]: I meant the class name itself. The list here only contains methods.

File: python/docs/source/reference/pyspark.sql/spark_session.rst [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    :toctree: api/
+
+The entry point to programming Spark with the Dataset and DataFrame API.
+To create a Spark session, you should use ``SparkSession.builder`` attribute.
+See also :class:`SparkSession`.
+
+.. autosummary::
+    :toctree: api/
+
+    SparkSession
```
Comment (HyukjinKwon) [inline python/docs/source/reference/pyspark.sql/spark_session.rst:34]: Here it contains `SparkSession` class but `python/docs/source/reference/pyspark.sql/window.rst` doesn't have `Window` class.

File: python/docs/source/reference/pyspark.sql/catalog.rst [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+=======
+Catalog
+=======
+
+.. currentmodule:: pyspark.sql
+
+.. autosummary::
+    :toctree: api/
+
+    Catalog.cacheTable
```
Comment (HyukjinKwon) [inline python/docs/source/reference/pyspark.sql/catalog.rst:28]: Here it doesn't have `Catalog` class but others like DataFrame or SparkSession have them.

File: python/docs/source/reference/pyspark.sql/functions.rst (hunk 2 above)
Comment (itholic) [reply python/docs/source/reference/pyspark.sql/functions.rst:28]: Let me make a sub-group

File: python/docs/source/reference/pyspark.sql/spark_session.rst (hunk 3 above)
Comment (itholic) [reply python/docs/source/reference/pyspark.sql/spark_session.rst:34]: Thanks. I think we can just remove `SparkSession` here, it's duplicated with `SparkSession` under "Core Classes"

File: python/docs/source/reference/pyspark.sql/catalog.rst (hunk 4 above)
Comment (itholic) [reply python/docs/source/reference/pyspark.sql/catalog.rst:28]: ditto as https://github.com/apache/spark/pull/36647#discussion_r880061516.

Let me remove `DataFrame`, in `dataframe.rst` too.

**Comment by itholic** [comment]
> Does the URL change? We might need to keep the mapping between old URL and new URL. If that's complicated, we can do it in a followup.

Yes, the URL has changed, let me find a way to keep the url as before.

File: python/docs/source/reference/pyspark.sql/core_classes.rst (hunk 1 above)
Comment (itholic) [reply python/docs/source/reference/pyspark.sql/core_classes.rst:37]: Got it. Just added

File: python/docs/source/reference/pyspark.sql/functions.rst [hunk 5]
```diff
@@ -99,174 +119,225 @@ Functions
...
     upper
-    var_pop
-    var_samp
-    variance
-    weekofyear
-    when
-    window
+
+
+UDF Functions
```
Comment (HyukjinKwon) [inline python/docs/source/reference/pyspark.sql/functions.rst:322]: UDF is short form of User Defined Function. So Let's just name it UDF

File: python/docs/source/reference/pyspark.sql/core_classes.rst [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    SparkSession
+    Catalog
+    DataFrame
+    Column
+    Observation
+    Row
+    GroupedData
+    PandasCogroupedOps
+    DataFrameNaFunctions
+    DataFrameStatFunctions
```
Comment (HyukjinKwon) [inline python/docs/source/reference/pyspark.sql/core_classes.rst:36]: Can we add DataFrameReader and DataFrameWriter?

**Comment by HyukjinKwon** [comment]
Looks pretty good otherwise

**Comment by itholic** [comment]
Can I fix the URL in the follow-up? Will create JIRA right after this PR is done.

**Comment by HyukjinKwon** [comment]
Yeah, let's file a separate JIRA.

File: python/docs/source/reference/pyspark.sql/io.rst [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+============
+Input/Output
+============
+.. currentmodule:: pyspark.sql
+
+.. autosummary::
+    :toctree: api/
+
+    DataFrameReader
```
Comment (HyukjinKwon) [inline python/docs/source/reference/pyspark.sql/io.rst:27]: can you remove this?

File: python/docs/source/reference/pyspark.sql/io.rst [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    DataFrameReader.json
+    DataFrameReader.load
+    DataFrameReader.option
+    DataFrameReader.options
+    DataFrameReader.orc
+    DataFrameReader.parquet
+    DataFrameReader.schema
+    DataFrameReader.table
+    DataFrameWriter.bucketBy
+    DataFrameWriter
```
Comment (HyukjinKwon) [inline python/docs/source/reference/pyspark.sql/io.rst:40]: ditto

File: python/docs/source/reference/pyspark.sql/configuration.rst [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+=============
+Configuration
+=============
+.. currentmodule:: pyspark.sql.conf
+
+.. autosummary::
+    :toctree: api/
+
+    RuntimeConfig
```
Comment (HyukjinKwon) [inline python/docs/source/reference/pyspark.sql/configuration.rst:27]: I think we should move this to Core class, and list up the methods under `RuntimeConfig`

File: python/docs/source/reference/pyspark.sql/io.rst [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    DataFrameReader.format
+    DataFrameReader.jdbc
+    DataFrameReader.json
+    DataFrameReader.load
+    DataFrameReader.option
+    DataFrameReader.options
+    DataFrameReader.orc
+    DataFrameReader.parquet
+    DataFrameReader.schema
+    DataFrameReader.table
```
Comment (HyukjinKwon) [inline python/docs/source/reference/pyspark.sql/io.rst:38]: DataFrameReader.text is missing

File: python/docs/source/reference/pyspark.sql/dataframe.rst [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+=========
+DataFrame
+=========
+
+.. currentmodule:: pyspark.sql
+
+.. autosummary::
+    :toctree: api/
+
+    DataFrame.agg
```
Comment (HyukjinKwon) [inline python/docs/source/reference/pyspark.sql/dataframe.rst:28]: to_pandas_on_spark
withColumns
withMetadata

are missing

**Review by HyukjinKwon — APPROVED** [review]
LGTM otherwise

**Comment by itholic** [comment]
Thanks for the review!

**Comment by itholic** [comment]
> Yeah, let's file a separate JIRA.

Just created: https://issues.apache.org/jira/browse/SPARK-39288

**Comment by itholic** [comment]
Just re-triggered the CI, but seems like previous failure is not related to my change

**Comment by HyukjinKwon** [comment]
All tests passed https://github.com/itholic/spark/runs/6602128754.

Merged to master.

**Pull request closed without merging**